frontend/src/metabase/setup/components/UserStep.jsx

```diff
@@ -105,18 +114,28 @@ export default class UserStep extends Component {
 
         this.props.setUserDetails({
             'nextStep': this.props.stepNumber + 1,
-            'details': {
-                'first_name': ReactDOM.findDOMNode(this.refs.firstName).value,
-                'last_name': ReactDOM.findDOMNode(this.refs.lastName).value,
-                'email': ReactDOM.findDOMNode(this.refs.email).value,
-                'password': ReactDOM.findDOMNode(this.refs.password).value,
-                'site_name': ReactDOM.findDOMNode(this.refs.siteName).value
-            }
+            'details': _.omit(fieldValues, "password_confirm")
         });
 
         MetabaseAnalytics.trackEvent('Setup', 'User Details Step');
     }
 
+    updateFieldValue = (fieldName, value) =>  {
+        this.setState({
+            fieldValues: {
+                ...this.state.fieldValues,
+                [fieldName]: value
+            }
+        }, this.validateForm);
+    }
+
+    onFirstNameChange = (e) => this.updateFieldValue("first_name", e.target.value)
+    onLastNameChange = (e) => this.updateFieldValue("last_name", e.target.value)
+    onEmailChange = (e) => this.updateFieldValue("email", e.target.value)
+    onPasswordChange = (e) => this.updateFieldValue("password", e.target.value)
+    onPasswordConfirmChange = (e) => this.updateFieldValue("password_confirm", e.target.value)
+    onSiteNameChange = (e) => this.updateFieldValue("site_name", e.target.value)
```

Comment:
I avoid arrow functions only for performance reasons. In your example `input` gets rendered again in every `render()` call because the arrow function always creates a new memory reference and React isn't able to figure out that it is still the same function as during the previous `render()` call.

That's more important for components that receive prop/state updates very often but I try to be pretty strict about following that rule (although in this case it doesn't have a noticeable performance impact).